Fix: #1941. Editorial alignments (#1973)
httpwg/http-extensions: draft-ietf-httpbis-digest-headers.md

This document defines the `Representation-Digest` request and response header
and trailer field ({{representation-digest}}) that contains a digest value

> computed by applying a hashing algorithm to *selected representation data*

Reviewer: I think we'll end up revisiting how best to draw attention to these terms. Emboldening doesn't seem like something the RFC editor would do but we can cross that bridge later on.

Author: Sure, I already filed an issue on kramdown for that ;)

Author: https://github.com/cabo/kramdown-rfc2629/issues <-

This document uses the Augmented BNF defined in [RFC5234] and updated by
[RFC7405].

> The terms Dictionary, List, Byte Sequence, sf-dictionary,
> sf-integer and sf-binary are imported from
> {{!STRUCTURED-FIELDS=RFC8941}}.

Reviewer: Borrowing MT's suggestion from https://github.com/httpwg/http-extensions/pull/1977/files#diff-4dc65e5537b9f2d8c77d359e6b7a032db79f294b488d2ae2b0554aa96f9b9258R133

```suggestion
This document uses the Boolean, Byte Sequence, 
Dictionary, Integer and List types from
{{!STRUCTURED-FIELDS=RFC8941}} along with
the sf-dictionary and sf-list ABNF rules.
```

The `Representation-Digest` HTTP field can be used in requests and responses to
communicate digests that are calculated using a hashing algorithm applied to

> the entire selected  *representation data* (see {{Section 8.1 of SEMANTICS}}).

Reviewer: above the attention is drawn to the term 

selected representation data

I can't recall if there's a reason this one is different by excluding selected.

several examples are provided in {{resource-representation}}.

When a message has no representation data it is still possible to assert that no

> representation data was sent computing the digest on an empty

Reviewer: old typo can be fixed too

```suggestion
representation data was sent by computing the digest on an empty
```

string (see {{usage-in-signatures}}).

`Representation-Digest` is a Structured Fields `Dictionary` (see {{Section 3.2 of
STRUCTURED-FIELDS}}) where:
* keys convey the hashing algorithm (see {{algorithms}})
  used to compute the digest

>   and have no parameters;

Reviewer: I don't follow this line, keys can't have parameters by design. If you mean that there are no parameters defined for Dictionary members, that probably needs to be done as a third bullet.

Author: I'll check the spec and fix

several examples are provided in {{resource-representation}}.

When a message has no representation data it is still possible to assert that no
representation data was sent by computing the digest on an empty
string (see {{usage-in-signatures}}).

`Representation-Digest` is a Structured Fields `Dictionary` (see {{Section 3.2 of
STRUCTURED-FIELDS}}) where:

> * members cannot have parameters;

Reviewer: This is something I sidestepped during the initial change to SF. The SF grammar allows parameters, so member-values *can* have them. The SF says

```
   Both Items and Inner Lists allow parameters as an extensibility
   mechanism; this means that values can later be extended to
   accommodate more information, if need be.  To preserve forward
   compatibility, field specifications are discouraged from defining the
   presence of an unrecognized parameter as an error condition.
```

When updating Digest we decided parameters could be a security problem. If we want prohibit the use of SF parameters, then we need some form of normative text to state that and probably justify why. I'd suggest doing that in a separate paragraph.

Reviewer: Furthermore, if we move the parameter discussion elsewhere, then we can get rid of the list format and use a prose style similar to what's already there.

Author: @LPardue I added a paragraph in the security consideration.

Representation-Digest   = sf-dictionary
~~~

where member-keys convey the hashing algorithm (see
{{algorithms}}) used to compute the digest
and have no parameters.
Member-values are the output of the digest calculation

> and their syntax is `sf-binary`.

Reviewer: How about we just move this to above the ABNF. So it reads

> `Representation-Digest` is a Structured Fields Dictionary (see {{Section 3.2 of
STRUCTURED-FIELDS}}). Member-keys convey ....
> 
> ~~~ abnf

> and their syntax is `sf-binary`.

Reviewer: Separately I'm not a big fan of saying "the syntax is sf-xyz" because it puts a lot of attention on syntax, which doesn't matter, and it doesn't put a normative requirement on the type, which is very important. 

How about "Member-values are the output of the digest calculation, the type MUST be Byte Sequence."?

This comment applies to other instances too.

Author: I  thought that syntax and type were tighly coupled, but if it's not the case (eg. a type can have different syntaxes)
it's ok.

As a follow up to your email, I think that a formal way to express relations between items - that is defining a "schema" for structured fields - would be great, but I suspect that it's not trivial. Moreover this could have sense only if fields are serialized differently from rfc8941.

  iYllu7BNNyealdVLvRwEmTHWXvJwew==:
~~~

> The `sf-dictionary` syntax, can

Author: @LPardue this text is now identical in both C-Digest and R-Digest. I guess if there's a way to avoid repetitions

  iYllu7BNNyealdVLvRwEmTHWXvJwew==:
~~~

> A recipient MAY ignore any or all `dict-member`s.

Author: @LPardue we could replace `dict-member`s with just "members" or "digests" (which is already used here https://github.com/httpwg/http-extensions/pull/1973/files#diff-de34152d297a9eb29b85b24962949ba2e309478a5c32e1d94fd4c3ee27b4584dL271 )

Author: "digest" is actually mentioned in Notational Conventions in relation to the values that we have in the fields... let's use it!

This allows the recipient to choose which hashing algorithm(s) to use for

> validation instead of verifying every received `dict-member`.

Reviewer: Good catch of representation-data-digest

``` suggestion
validation instead of verifying every `dict-member`.
```

> validation instead of verifying every received `dict-member`.

Reviewer: Good catch of representation-data-digest

``` suggestion
validation instead of verifying every `dict-member`.
```

The `Content-Digest` HTTP field can be used in requests and responses to
communicate digests that are calculated using a hashing algorithm applied to
the actual message content (see {{Section 6.4 of SEMANTICS}}). It is a
Structured Fields Dictionary (see {{Section 3.2 of STRUCTURED-FIELDS}})
where:
* keys convey the hashing algorithm (see {{algorithms}})
  used to compute the digest
  and have no parameters;
* values MUST be `Byte Sequences`

>   containing the output of the digest calculation.

Reviewer: Same points here, as for Representation digest

the actual message content (see {{Section 6.4 of SEMANTICS}}). It is a
Structured Fields Dictionary (see {{Section 3.2 of STRUCTURED-FIELDS}})
where:

> * members MUST NOT have parameters (see {{sec-agility}});

Reviewer: Furthermore, prohibiting parameters suffers can lead to the same fate of validation DoS.

So the text that discusses normative behaviour of parameters has to be more nuanced than a single line.

> * members MUST NOT have parameters (see {{sec-agility}});

Reviewer: Understood but it's likely that such an editorial comment would come up during our later review anyway. I'll take a go and implementing my suggestion.

> * members MUST NOT have parameters (see {{sec-agility}});

Author: > I'll take a go and implementing my suggestion. 

:rocket:

> * members MUST NOT have parameters (see {{sec-agility}});

Author: Parameters will be managed in another PR

~~~

Dictionary members convey hashing algorithm preferences (see {{algorithms}}).
Member-keys convey the hashing algorithm (see {{algorithms}}),
member-values convey an ascending relative weighted preference

> and their syntax is `sf-integer` in the range 0 to 10 inclusive. 1 is the

Author: Removed duplicated text.

the `Content-Digest` field.

`Want-Representation-Digest` and `Want-Content-Digest` are Structured Fields
List (see {{Section 3.2 of STRUCTURED-FIELDS}}) where:

> * members convey hashing algorithm preferences (see {{algorithms}});

Reviewer: we already state that above, so this seems duplicative
```suggestion
```

the `Content-Digest` field.

`Want-Representation-Digest` and `Want-Content-Digest` are Structured Fields

> List (see {{Section 3.2 of STRUCTURED-FIELDS}}) where:

Reviewer: ```suggestion
Dictionary (see {{Section 3.2 of STRUCTURED-FIELDS}}) where:
```

the `Content-Digest` field.

`Want-Representation-Digest` and `Want-Content-Digest` are Structured Fields
List (see {{Section 3.2 of STRUCTURED-FIELDS}}) where:

> * members cannot have parameters;

Reviewer: This case is a bit different, what is good justification to prohibit parameters on want-* fields?

Author: Ok.

## HTTP Messages Are Not Protected In Full {#sec-limitations}

This document specifies a data integrity mechanism that protects HTTP
*representation data* or content, but not HTTP header and trailer fields, from
certain kinds of corruption.

Integrity fields are not intended to be a general protection against malicious tampering with
HTTP messages. This can be achieved by combining it with other approaches such
as transport-layer security or digital signatures.

## End-to-End Integrity

Integrity fields can help detect  *representation data* or content modification due to implementation errors,
undesired "transforming proxies" (see {{Section 7.7 of SEMANTICS}})
or other actions as the data passes across multiple hops or system boundaries.
Even a simple mechanism for end-to-end  *representation data* integrity is valuable

> because a *user agent* can validate that resource retrieval succeeded before handing off to a

Reviewer: The grammer fix is good but there doesn't seem to be any need to treat user agent as special by adding '*'

Using signatures to protect the checksum of an empty representation
allows receiving endpoints to detect if an eventual payload has been stripped or added.

Any mangling of digest fields, including de-duplication of `dict-member`s
or combining different field values (see {{Section 5.2 of SEMANTICS}})

> might affect signature validation.

Reviewer: Dictionaries already describe how to treat duplicate entries. So we probably need a separate issue to fix up this paragraph.

Author: Once a field is signed, even a compliant mangling invalidates the signature. I see this line as a simple advice to the users.

such as preferring "standard" algorithms over "insecure" ones.
Transition from weak algorithms is supported
by negotiation of hashing algorithm using `Want-Representation-Digest` or `Want-Content-Digest` (see {{want-fields}})

> or by sending multiple digest calculations from which the receiver chooses.

Reviewer: or simply
```suggestion
or by sending multiple digests from which the receiver chooses.
```

--- back

> # Resource Representation and Representation Data {#resource-representation}

Reviewer: This is a weird title now that we removed the term `Representation-Data`. Maybe there's a better alternative?

Author: We removed `representation-data-digest`. Representation data is the concept defined in SEMANTICS.

## Since draft-ietf-httpbis-digest-headers-01
{:numbered="false"}

> * Digest of error responses is computed on the error representation data #1004

Reviewer: I don't think we need to worry about updating terms in the changelog. That can quickly become an impossible task.

Author: it was just a s//g ;)
